[OpenCL C 3.0] Extended description of optional functionality (#330). (#387)
KhronosGroup/OpenCL-Docs: OpenCL_C.txt

Please refer to the C99 specification for a detailed description of the
language grammar.

[[optional-functionality]]
== Optional functionality

Some language functionality is optional and will not be supported by all
devices. Such functionality is represented by optional language features or
language extensions. Support of optional functionality in OpenCL C is indicated
by the presence of special predefined macros.

[[features]]
=== Features

Optional language features are described in this document. They are optional
from OpenCL C 3.0 and therefore are not supported by all implementations. When
an OpenCL C 3.0 optional feature is supported, an associated __feature test
macro__ will be predefined.

The following table describes OpenCL C 3.0 features and their meaning. The

> naming convension for the feature macros is `+__opencl_c_<name>+`.

Reviewer: ```suggestion
naming convention for the feature macros is `+__opencl_c_<name>+`.
```

Please refer to the C99 specification for a detailed description of the
language grammar.

[NOTE]
--
The OpenCL 3.0 C programming language described in this specification is provisional.
It has been Ratified under the Khronos Intellectual Property Framework and is being made publicly available to enable review and feedback from the community.

If you have feedback please create an issue on https://github.com/KhronosGroup/OpenCL-Docs/

> --

Reviewer 2: This section has been removed from the latest specifications.

Author: I have rebased now, so this section is removed. Thanks!

[[optional-functionality]]
== Optional functionality

> Some language functionality is optional that will not be supported by all

Reviewer: ```suggestion
Some language functionality is optional and will not be supported by all
```

Or some other change to make this sentence parse.

devices. Such functionality is represented by optional language features or
language extensions. Support of optional functionality in OpenCL C is indicated
by the presence of special predefined macros.

[[features]]
=== Features

Optional language features are described in this document. They are optional

> from OpenCL 3.0 and therefore are not supported by all devices. When an OpenCL

Reviewer 2: I have a few minor suggestions for this sentence:

1. I've been trying to clearly distinguish between **OpenCL 3.0** and **OpenCL C 3.0**, especially since **OpenCL C** is being used by other APIs (e.g. `clspv`).  With that in mind, I think this sentence should be changed to "OpenCL C 3.0" instead of "OpenCL 3.0".  Do you agree?

2. If so, should this sentence also be updated to indicate that these optional features may not be supported by all "compilers" or "implementations" instead of by all "devices"?

Author: Both are addressed in the last commits. Thanks!

C 3.0 optional feature is supported, an associated __feature test macro__ will
be predefined.

The following table describes OpenCL C 3.0 features and their meaning. The
naming convension for the feature macros is `+__opencl_c_<name>+`.

Feature macro identifiers are used as names of features in this document.

[[table-optional-lang-features]]

> .Optional features in OpenCL 3.0 and their predefined macros.

Reviewer: ```suggestion
.Optional features in OpenCL C 3.0 and their predefined macros.
```

To match Ben's OpenCL C point.

object in a kernel.

OpenCL C compilers that define the feature macro
`+__opencl_c_read_write_images+` must also define the feature macro
`+__opencl_c_images+`.

| `+__opencl_c_subgroups+`
| The OpenCL C compiler supports built-in functions operating on sub-groupings
of work-items.

| `+__opencl_c_work_group_collective_functions+`
| The OpenCL C compiler supports built-in functions that perform collective
operations across a work-group.

|====

In OpenCL C 3.0, feature macros must expand to the value `1` if the feature
macro is defined by the OpenCL C compiler. A feature macro must not be
defined if the feature is not supported by the OpenCL C compiler. A feature
macro may expand to a different value in the future, but if this occurs the
value of the feature macro must compare greater than the prior value of the
feature macro.

[[extensions]]
=== Extensions

Optional functionality that is not defined in this document is referred to
as extensions. Extensions are described in
<<opencl-extension-spec,the OpenCL Extension Specification>>.

[NOTE]
--
Prior to OpenCL 3.0 some optional features described in this document were

> referred to as optional core language extensions. Their presence could be

Reviewer: ```suggestion
referred to as optional core features. Their presence could be
```

Or maybe `optional core language features`.   The main examples of such language are in [appendix A of the extension spec](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#_extensions_promoted_to_core_features).

Author: Btw is the Appendix A not part of the spec document? I was thinking to add a reference but not sure how...

> indicated by the predefined extension macros. If any of the features has been

Reviewer: > indicated by the predefined extension macros.

Is this intended to cover features such as `__IMAGE_SUPPORT__`?  Because that was never an extension.

Author: Ok, I didn't mean to include originally it but we can do if you think it is needed. The reason I didn't include it is mainly because it is documented in Predefined macros section. So technically both `__IMAGE_SUPPORT__` and `__opencl_c_images` are documented in OpenCL C explicitly.

Reviewer: Yes, you're right, the text is fine as-is.

by OpenCL are reserved for future use.

The predefined identifier `+__func__+` is available.

> In OpenCL C 3.0 there are a number of optional predefined macro indicating

Reviewer: ```suggestion
In OpenCL C 3.0 there are a number of optional predefined macros indicating
```

by OpenCL are reserved for future use.

The predefined identifier `+__func__+` is available.

In OpenCL C 3.0 there are a number of optional predefined macro indicating

> optional language features. Such macros are listed in
> <<table-optional-lang-features>>.

Reviewer: ```suggestion
optional language features. Such macros are listed in the
<<table-optional-lang-features, optional features in OpenCL C 3.0 table>>.
```

Not strictly required, but as is it ends up with a `..` at the end of the sentence.
